### PR TITLE
Set allowsBackgroundLocationUpdates as necessary when using location …

### PR DESCRIPTION
**Pull request by briancalvium**
Set the flag 'allowsBackgroundLocationUpdates' to YES if the location background mode is enabled and the 'Always Allow Location' key is set in the App Plist.

**motivation**

We found that on iOS 9.x, the allowsBackgroundLocationUpdates flag must be set on location manager in order to receive updates when the app is in the background (seems to affect actual device only).

**Test plan (required)**
Example app using the forked branch [here](https://github.com/briancalvium/react-native-geolocation-pr-example). Run this example through XCode on a device with 9.x, observe that location update logs continue to appear when the app is in the background. Switch to react native 0.32.0, observe that location update logs stop once the app is in the background.

### CODE REVIEW

**Comment by facebook-github-bot-0** [comment]
By analyzing the blame information on this pull request, we identified @nicklockwood and @majak to be potential reviewers.

**Comment by facebook-github-bot-2** [comment]
@briancalvium updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9717/files/e3ba14a89c835c5e2a1fbafce6cc7b9e18e2ec08..9525ad191f1941934dd3cec574702c30a35e74d9)

**Comment by facebook-github-bot-2** [comment]
@briancalvium updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9717/files/9525ad191f1941934dd3cec574702c30a35e74d9..12a1a736197db8098d37a28866af353237ab169e)

**Comment by Kerumen** [comment]
Agree, also need this!

**Comment by mkonicek** [comment]
What does it mean to "enable geolocation in the background"? React Native apps don't run in the background. Does it mean the location will keep getting updated and when the app is brought into foreground the app will see up-to-date location?

Why not only update the location when the app is in foreground?

**Comment by briancalvium** [comment]
@mkonicek Interesting - I have been able to console.log and perform other actions in JS with the app in the background with this flag enabled (and the location BG mode in the plist). For instance, open the app, start updating location, press the home button, and continue to see location update logs from a component. Should we perhaps not depend on this behavior in real-world conditions?

Ultimately, we want to geofence and notify the user either by local notification or audio when the phone is in the user's pocket, like in an audio tour. I've been working under the assumption that the location background mode wakes up the app 'enough' to run JS when an update comes in, which has seemed to be the case so far. (we are using [react-native-location](https://github.com/timfpark/react-native-location) for now, which exposes this property, and claims it works as well)

**Comment by mkonicek** [comment]
@briancalvium Wow, I've learned something new about React Native on iOS :) Thanks! 👍

File: Libraries/Geolocation/RCTLocationObserver.m [hunk 1]
```diff
@@ -144,6 +144,14 @@ - (void)beginLocationUpdatesWithDesiredAccuracy:(CLLocationAccuracy)desiredAccur
...
     [_locationManager respondsToSelector:@selector(requestAlwaysAuthorization)]) {
     [_locationManager requestAlwaysAuthorization];
+
+    // On iOS 9+ we also need to enable background updates
+    NSArray* backgroundModes  = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"UIBackgroundModes"];
+    if(backgroundModes && [backgroundModes containsObject:@"location"]) {
+      if([_locationManager respondsToSelector:@selector(setAllowsBackgroundLocationUpdates:)]) {
+        [_locationManager setAllowsBackgroundLocationUpdates:YES];
+      }
+    }
```
Comment (mkonicek) [inline Libraries/Geolocation/RCTLocationObserver.m]: Will this always enable background updates when the app starts listening for location updates?

Or will it only be enabled when configured as described in `Geolocation.js`?

Sorry for the iOS n00b questions :)

**Comment by mkonicek** [comment]
https://github.com/timfpark/react-native-location looks cool, like the docs!

File: Libraries/Geolocation/RCTLocationObserver.m (hunk 1 above)
Comment (Kerumen) [inline Libraries/Geolocation/RCTLocationObserver.m]: You will need to configure the `Info.plist` file to enable it.

From the [Apple iOS docs](https://developer.apple.com/library/ios/documentation/CoreLocation/Reference/CLLocationManager_Class/#//apple_ref/occ/instp/CLLocationManager/allowsBackgroundLocationUpdates):

> The default value of this property is `NO`. Setting the value to `YES` but omitting the `UIBackgroundModes` key and `location` value in your app’s `Info.plist` file is a programmer error.

Comment (briancalvium) [inline Libraries/Geolocation/RCTLocationObserver.m]: There is also a case to be made to expose this property to JS, as there are times when you might want to toggle the background updates dynamically, but it didn't seem to fit in with the Mozilla geolocation API, so I thought this fit best with what is there already.

**Pull request closed without merging**